### ISSUE TITLE
Request help on how to stop the gradient flow in nccl_allreduce?

### ISSUE THREAD

**Issue by Remember2018** [tensorpack/tensorpack#1355]
Hi Yuxin, in multi-GPU setting, to compute one loss, I need to use the `nccl` ops to reduce-sum one variable. Some code pieces are as follows:

    one_scalar_sum= tf.reduce_sum(one_scalar, name='one_scalar_sum')
    one_scalar_sum= tf.stop_gradient(one_scalar_sum)

    if cfg.TRAIN.NUM_GPUS > 1:
        shared_name = 'one_scalar_sum'
        one_scalar_sum_avg= gen_nccl_ops.nccl_all_reduce(
                    input=one_scalar_sum,
                    reduction='sum',
                    num_devices=cfg.TRAIN.NUM_GPUS,
                    shared_name=shared_name) * (1. / cfg.TRAIN.NUM_GPUS)
    else:
        one_scalar_sum_avg= one_scalar_sum

    one_scalar_sum_avg= tf.stop_gradient(one_scalar_sum_avg)
    loss = tf.reduce_sum(loss) / one_scalar_sum_avg

As shown in the code above, I have stopped the gradient flow to back propagate to the previous tensors. It is expected that no gradients were back flowed to the previous tensors. 

[![tracing][1]][1]

But when I used the graph profiler to check the execution time of these ops, it seems there are some gradient flow operations were made in the graph. I have checked the source code of `graph_builder`.

https://github.com/tensorpack/tensorpack/blob/7e923bfe7fdb97f5e5f9c9ef0df9967658fa5b5f/tensorpack/graph_builder/utils.py#L132

It seems these ops for gradient flow were automatically generated from the Tensorpack API. So in this case, how to stop those gradient flows for the `nccl` ops?   From my understanding of Tensorflow, if I have stopped the gradient in one tensor, then the gradient flow will not be back propagated from the previous tensors before it, is this correct?

I know maybe this question is not related to `Tensorpack`, but I don't know whom I could request for help. If you are convenient, could you please give some advice? Thanks.

  [1]: https://i.stack.imgur.com/3gQIT.png

**Comment by ppwwyyxx** [comment]
gradient will not propagate from `one_scalar_sum_avg` but will still propagate from `loss` because you did not stop gradient on `loss`.

**Comment by Remember2018** [comment]
Yes, from the last line:

    final_loss = tf.reduce_sum(loss) / one_scalar_sum_avg

It is expected that the error from the `final_loss` would just back to the `loss` branch, but not back to the `one_scalar_sum_avg` branch. So I stop the gradient flow on the `one_scalar_sum_avg`. But if the error flow is stopped in `loss`, then the parameters will not be updated by the gradient descent. Is my understanding correct?

In addition, I have another question on the `backbone` in `FasterRCNN` module. In the 

https://github.com/tensorpack/tensorpack/blob/7e923bfe7fdb97f5e5f9c9ef0df9967658fa5b5f/examples/FasterRCNN/modeling/backbone.py#L66

![CA3E977F-9CE8-4dc3-93E5-2FD0BFEAF863](https://user-images.githubusercontent.com/43327429/67620622-7cad4580-f83b-11e9-98d7-2d7a6d1fc307.png)

But in other backbone for `ResNet50`,  there are some `bias`s:

![0E71DA90-52B5-457e-BB53-7B1350897BDE](https://user-images.githubusercontent.com/43327429/67620650-bda55a00-f83b-11e9-844a-9d9a338115c8.png)

And from the code:

https://github.com/tianzhi0549/FCOS/blob/b516eb54dc069d9c0fadb0d51eeead7dd1bb4910/fcos_core/solver/build.py#L15

there are some configurations on the `learning_rate` for the `bias`. But these configurations are removed from the `FasterRCNN` module. I must miss something, could you please give some advice? Thanks for your time.

**Comment by ppwwyyxx** [comment]
> Is my understanding correct?

yes

> But in other backbone for ResNet50, there are some biass:

It's a bug in those models and they are not used.

> there are some configurations on the learning_rate for the bias

they do not matter.

**Comment by Remember2018** [comment]
Hi Yuxin, thanks a lot for your kind comments! 
In regard to the gradients of `nccl` ops, I followed your great suggestions noted in the `Performance tuning` section of the documentation of `Tensorpack`. I found those  gradients are not expected in the above case. I'm not sure if those branches would run in the training procedure. If they are executed but they have no contribution to the parameter update, then they can be cut or removed in the graph, and then speed up the training.
